pandas/src/inference.pyx

```diff
@@ -712,165 +775,120 @@ def maybe_convert_numeric(ndarray[object] values, set na_values,
     cdef:
         int status, maybe_int
         Py_ssize_t i, n = values.size
+        Seen seen = Seen(coerce_numeric);
         ndarray[float64_t] floats = np.empty(n, dtype='f8')
         ndarray[complex128_t] complexes = np.empty(n, dtype='c16')
         ndarray[int64_t] ints = np.empty(n, dtype='i8')
         ndarray[uint64_t] uints = np.empty(n, dtype='u8')
         ndarray[uint8_t] bools = np.empty(n, dtype='u1')
-        bint seen_null = False
-        bint seen_uint = False
-        bint seen_sint = False
-        bint seen_float = False
-        bint seen_complex = False
-        bint seen_int = False
-        bint seen_bool = False
         object val
         float64_t fval
 
     for i in range(n):
         val = values[i]
 
         if val.__hash__ is not None and val in na_values:
-            seen_null = True
-            if _check_uint64_nan(seen_uint, seen_null,
-                                 coerce_numeric):
-                return values
-
+            seen.seen_null = True
```

Comment:
I would make each of these cases where you are currently setting a method of Seen instead

e.g.

```
def saw_float(self):
    self.seen_nul = True
    self.seen_float = True
```